lib/farscape/client/http_client.rb

```diff
@@ -0,0 +1,99 @@
+require 'faraday'
+require 'farscape/client/base_client'
+require 'farscape/plugins'
+require 'farscape/errors'
+
+module Farscape
+  class Agent
+    class HTTPClient < BaseClient
+
+      # The Faraday connection instance.
+      attr_reader :connection
+      attr_reader :Plugins
+
+      def initialize(agent)
+        @connection = Faraday.new do |builder|
+          builder.request :url_encoded
+          agent.middleware_stack.each do |middleware|
+            if middleware.key?(:config)
+              config = middleware[:config]
+              if config.is_a?(Array)
+                builder.use(middleware[:class], *config)
+              else
+                builder.use(middleware[:class], config)
+              end
+            else
+              builder.use(middleware[:class])
+            end
+          end
+
+          builder.adapter faraday_adapter
+        end
+      end
+
+      # Override this in a subclass to create clients with custom Faraday adapters
+      def faraday_adapter
+        Faraday.default_adapter
+      end
+
+      ##
+      # Makes a Faraday request given the specified options
+      #
+      # @options [Hash] The hash of Faraday options passed to the request, including url, method,
+      #  params, body, and headers.
+      # @return [Faraday::Response] The response object resulting from the Faraday call
+      def invoke(options = {})
+        defaults = { method:  'get'}
+        options = defaults.merge(options)
+
+        connection.send(options[:method].to_s.downcase) do |req|
+          req.url options[:url]
+          req.body = options[:body] if options.has_key?(:body)
+          options[:params].each { |k,v| req.params[k] = v } if options.has_key?(:params)
+          options[:headers].each { |k,v| req.headers[k] = v } if options.has_key?(:headers)
+        end
+      end
+
+      def interface_methods
+        {
+          idempotent: ['PUT', 'DELETE'],
+          unsafe: ['POST', 'PATCH'], # http://tools.ietf.org/html/rfc5789
+          safe: ['GET', 'HEAD', 'OPTIONS', 'TRACE', 'CONNECT']
+        }
+      end
+
+      def dispatch_error(response)
+        errors = Farscape::Exceptions
+        http_code = {
```

Comment:
why isn't this just a constant?  I'm pretty sure that we don't need to instantiate this hash every time farscape dispatches an error? 

Oh, better yet, don't declare it as an hash.  I see that the whole point of this code is that you're avoiding writing a simple ruby case-statement.  Just use a case statement.   No need to even declare a hash.